python_modules/libraries/dagster-components/dagster_components/resolved/model.py

```diff
@@ -153,16 +218,43 @@ def from_spec(spec: type[ResolvedKwargs], target_type: type):
 
     @staticmethod
     def from_annotation(annotation: Any, field_name: str) -> "FieldResolver":
-        if get_origin(annotation) is Annotated:
-            args = get_args(annotation)
+        if _is_scalar(annotation):
+            return FieldResolver.from_model(
+                lambda context, model: context.resolve_value(getattr(model, field_name))
+            )
+
+        origin = get_origin(annotation)
+        args = get_args(annotation)
+
+        if origin is Annotated:
             resolver = next((arg for arg in args if isinstance(arg, FieldResolver)), None)
             if resolver:
                 return resolver
 
-            check.failed(f"Could not find resolver on annotation {field_name}")
-
-        return FieldResolver.from_model(
-            lambda context, model: context.resolve_value(getattr(model, field_name))
+        if origin in (Union, UnionType) and len(args) == 2:
+            left_t, right_t = args
+            if right_t is type(None):
+                resolved_from_cls = _get_resolved_from_cls(left_t)
+                if resolved_from_cls:
+                    return FieldResolver(resolved_from_cls.from_optional)
+
+                elif get_origin(left_t) in (Sequence, tuple, list):
+                    resolved_from_cls = _get_resolved_from_cls(get_args(left_t)[0])
+                    if resolved_from_cls:
+                        return FieldResolver(resolved_from_cls.from_optional_seq)
+
+        elif origin in (Sequence, tuple, list):
+            resolved_from_cls = _get_resolved_from_cls(args[0])
+
+            if resolved_from_cls:
+                return FieldResolver(resolved_from_cls.from_seq)
+
+        check.failed(
+            f"Could not derive resolver for annotation {field_name}: {annotation}.\n"
+            "Field types are expected to contain:\n"
+            "* basic serializable types such as str, int, float, bool, list, etc\n"
+            "* ResolvedFrom subclasses wrapped with Resolved[]\n"
```

Comment:
there might be better ways to word it that are not as egregious, but I feel like this would be an annoying error to hit 

"why do i have to mark it `Resolved[]` if it already knows its a `ResolvedFrom` subclass" 

compare with https://github.com/dagster-io/dagster/pull/28366/files#r1988087666

Reply:
I much prefer leaving `Resolved[]` out